packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts

```diff
@@ -189,22 +190,21 @@ export function useColumns(columns: Columns, apiRef: ApiRef): void {
         !!options.checkboxSelection,
         logger,
       );
+      const updatedCols = updateColumnsWidth(hydratedColumns, gridState.viewportSizes.width);
 
       updateState({
-        all: hydratedColumns.map((col) => col.field),
-        lookup: toLookup(logger, hydratedColumns),
+        all: updatedCols.map((col) => col.field),
+        lookup: toLookup(logger, updatedCols),
       });
     } else {
       updateState(getInitialColumnsState());
     }
-  }, [logger, columns, options.columnTypes, options.checkboxSelection, updateState]);
-
-  React.useEffect(() => {
-    logger.debug(`Columns gridState.viewportSizes.width, changed ${gridState.viewportSizes.width}`);
-    // Avoid dependency on gridState as I only want to update cols when viewport size changed.
-    const currentColumns = allColumnsSelector(apiRef.current.getState<GridState>());
-
-    const updatedCols = updateColumnsWidth(currentColumns, gridState.viewportSizes.width);
-    apiRef.current.updateColumns(updatedCols);
-  }, [apiRef, gridState.viewportSizes.width, logger]);
+  }, [
+    logger,
+    columns,
+    options.columnTypes,
+    options.checkboxSelection,
+    gridState.viewportSizes.width,
```

Comment:
If you gather the 2 useEffect, you add a dependency on viewport size which means that if you're going to reset the columns every time you resize. Hence why I did 2 separate effects.
Try reorder, resize columns, and then display debugger => col reset to initial state
Just restore the previous effect, so we just update the width when the screen resize

Reply:
Should we also add a test case for the resize? Shouldn't the CI have failed?

Reply:
Unfortunately the width needs to be updated when not only when the screen size changes but also when the data set does, thats why I combined them. If you want I can separate them into to useEffects but the one that updates the width will need to have almost the same dependencies.

Reply:
@DanailH What do you think of breaking down this effort into two chunks? Is it accurate that we can solve the `hide` aspect independently of the rerender issue?

Reply:
The problem is that the root cause for both is the same and that these are only 2 of the possible ways this issue can occur. Basically the way that it is refactored now only calculates the width of the columns on initial load and on screen resize which is wrong because if the columns change (hide, show etc) or the rows disappear the columns hydrate function kicks in and defaults the width of the columns back to 100. Basically what I mean is that the check for the flex should be part of the normal column hydration process.

Reply:
@DanailH I might be missing something because it looks like we can close #708 by only adding the `column.hide` logic and test case for https://github.com/mui-org/material-ui-x/issues/708#issuecomment-748465868.

Reply:
@oliviertassinari yes that will solve it on the initial render but if you hide/show the column dynamically it will still break. The thing is that if we do it that way then this won't solve the reported [issue in the description ](https://github.com/mui-org/material-ui-x/issues/708#issue-759726684)

Reply:
I think that we should split the efforts as much as we can and fix bugs as isolated as possible. We could have two test cases here. Let's say we fix the re-rendering bug; #716, will we solve #708? The answer is no, it's not enough.

Reply:
Actually, we probably need three test cases, not two. The first one for #716, the second one for #708, and the thrid one to fail with the current changes. We should avoid having Damien to report the regression again in the future. It's especially important as reproducing it requires an interaction:

![bug](https://user-images.githubusercontent.com/3165635/102766852-0d6bac80-437f-11eb-91c1-a481c45145c0.gif)